Memory improvement: removing most of Long.valueOf()

Armaxis opened the pull request:
## Intro
In the past we have discovered that there are a lot of boxings of primitive `long` into `Long` that take some CPU time and trigger additional allocations, adding more work for GC.
We've been addressing this issue for a while:
- #1850 removed a lot of `Long.valueOf()` usages + extra memory improvements
- #1853 is going to replace `LruCache<Long, Object>` with `LongLruCache<Object>`
- #1750 was removing some of the `Long.valueOf()` usages (current PR is based on it)

With this PR we're removing **~80%** of `Long` allocations. Detailed benchmarks are below

## Changes
- Removed a lot of hidden calls to Long.valueOf()
  - From `Pair<Long, Object>`
  - From `Set<Long>` / `HashSet<Long>`
  - From `asObjectId` wrapping with `Long?`
- Added `remove` method to `LongScatterSet`
- Added `LongPair<B>` and `LongLongPair`

## Benchmarks

### Performance
Measured with [benchmarkCode](https://github.com/square/leakcanary/blob/1e6b9dde5d7217a5c154528addba097fcbf873f3/leakcanary-android-instrumentation/src/androidTest/java/leakcanary/Benchmark.kt#L32) on [ProfiledTest](https://github.com/square/leakcanary/blob/5983eae9a1dbd96f0de5cd5e2601cd3c563f978f/leakcanary-android-instrumentation/src/androidTest/java/leakcanary/ProfiledTest.kt) on non-rooted **Pixel 3** (Android 10), **30 iterations**.

**Without optimizations:**
Durations (ms): median **10523.5**, min 10383, max 11252

**With optimizations:**
Durations (ms): median **9931.0**, min 9786, max 10165

**Results**: improvement of **592.5 ms** which is **5.6%**.

### Method Tracing
Acquired with [runWithMethodTracing](https://github.com/square/leakcanary/blob/5983eae9a1dbd96f0de5cd5e2601cd3c563f978f/leakcanary-android-instrumentation/src/androidTest/java/leakcanary/Profiler.kt#L67) on [ProfiledTest](https://github.com/square/leakcanary/blob/5983eae9a1dbd96f0de5cd5e2601cd3c563f978f/leakcanary-android-instrumentation/src/androidTest/java/leakcanary/ProfiledTest.kt) on non-rooted **Pixel 3** (Android 10).

Method tracing files (with and without optimizations): [Traces.zip](https://github.com/square/leakcanary/files/4808634/Traces.zip)

_Tip: open another tab with this PR to compare before/after screenshots._

Bottom Up View, sorted by **Self %**

**Without Optimization**
![image](https://user-images.githubusercontent.com/2152942/85212721-2f766a80-b31b-11ea-8a0f-d92e1064ba19.png)
![image](https://user-images.githubusercontent.com/2152942/85212723-3309f180-b31b-11ea-9765-1a786c2f92fc.png)

**With Optimization**
![image](https://user-images.githubusercontent.com/2152942/85212711-1a99d700-b31b-11ea-9f7e-937cc08eb32c.png)
![image](https://user-images.githubusercontent.com/2152942/85212716-25546c00-b31b-11ea-8c22-7b2826fbcb8b.png)

**Results**: without optimization, `Long.valueOf()` accounted for **2.55% (0.8% self)** of execution time. With optimization - it's **0.67% (0.28% self)**.

### Allocations
Acquired manually by attaching Profiler before running [ProfiledTest](https://github.com/square/leakcanary/blob/5983eae9a1dbd96f0de5cd5e2601cd3c563f978f/leakcanary-android-instrumentation/src/androidTest/java/leakcanary/ProfiledTest.kt).

**Without Optimization**

**Sorted by Allocations**
![image](https://user-images.githubusercontent.com/2152942/85212777-e246c880-b31b-11ea-9f73-61cde8c1e9f7.png)

**Sorted by Shallow size**
![image](https://user-images.githubusercontent.com/2152942/85212785-ef63b780-b31b-11ea-8007-7bbd3e0ad9eb.png)

**Callstack of `analyzeGraph()`**
![image](https://user-images.githubusercontent.com/2152942/85212796-00142d80-b31c-11ea-8f9a-b171dd356b1e.png)

**With Optimization**

**Sorted by Allocations**
![image](https://user-images.githubusercontent.com/2152942/85212803-0e624980-b31c-11ea-8b90-482994f8e12b.png)

**Sorted by Shallow size**
![image](https://user-images.githubusercontent.com/2152942/85212807-15895780-b31c-11ea-895d-211bb816da4b.png)

**Callstack of `analyzeGraph()`**
![image](https://user-images.githubusercontent.com/2152942/85212814-1cb06580-b31c-11ea-9884-45e71b02a66d.png)

**Results**: number of `Long` allocations went down by **~500 000** which is **~80%**. Shallow size is unchanged (as expected). `Pair` allocations replaced with `LongPair` and `LongLongPair`. Total number of allocations went down by **~ 17%**.

Armaxis commented on shark/src/main/java/shark/internal/PathFinder.kt:
```diff
@@ -382,7 +385,7 @@ internal class PathFinder(
       }
 
-      val objectId = staticField.value.asObjectId!!
+      val objectId = (staticField.value.holder as ReferenceHolder).value
```
As an alternative I could've added `staticField.value.asObjectIdOrThrow` method, but that would be changing the public API. 
We were using `asObjectId!!` which means that we are sure that the holder inside value is a `ReferenceHolder`. This implies that we know what's going on and are ok with replacing this assumption on equivalent of `staticField.value.holder as ReferenceHolder`.

Decompiled Java.
With optimization:
```
ValueHolder var12 = staticField.getValue().getHolder();
if (var12 == null) {
   throw new TypeCastException("null cannot be cast to non-null type shark.ValueHolder.ReferenceHolder");
}
long objectId = ((ReferenceHolder)var12).getValue();
```

Without optimization:
```
Long var12 = staticField.getValue().getAsObjectId();
if (var12 == null) {
   Intrinsics.throwNpe();
}
long objectId = var12;
```

pyricau replied:
Add comment?

Armaxis commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+
+/**
+ * Replacement to a traditional Pair<Long, Object> that doesn't box long.
+ * Class is purposefully not made as a `data class` to decrease memory footprint of the object.
```
Based on allocation view `LongPair` takes 24 bytes, while regular `Pair` is 16. Makes sense, since `LongPair` contains `long` in it (8 bytes), while `Pair` references another `Long` object.

pyricau commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  override fun toString(): String {
+    return "LongPair(first=$first, second=$second)"
```
- [ ] Change to LongLongPair

pyricau commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+ * Replacement to a traditional Pair<Long, Object> that doesn't box long.
+ * Class is purposefully not made as a `data class` to decrease memory footprint of the object.
+ */
+class LongPair<out B>(
```
class should be internal, you don't want to pollute everyone's classpath

Armaxis replied:
`PathFinder` (from `shark` module) uses `LongLongScatterMap` (from `shark-graph`) which uses `LongLongPair` (also from `shark-graph`). Can't make `LongLongPair` internal :(

pyricau replied:
Copy paste will do until we have internal across modules.

pyricau commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+ * Replacement to a traditional Pair<Long, Long> that doesn't box long.
+ * Class is purposefully not made as a `data class` to decrease memory footprint of the object.
+ */
+class LongLongPair(
```
class should be internal

Armaxis replied:
See another comment on why I can't

pyricau commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+infix fun <B> Long.longTo(that: B): LongPair<B> = LongPair(this, that)
```
Rename both methods to `to` ? Then it's a matter of using the right import? Thoughts?

pyricau commented on shark-graph/src/main/java/shark/internal/hppc/LongScatterSet.kt:
```diff
@@ -113,6 +113,60 @@ class LongScatterSet {
+      }
+      false
+    }
+  }
```
yeah, I tried to remove what I didn't need which might have yielded other optimizations / inlining, so you'll want to write tests for this ;)

pyricau replied:
I don't understand why you're adding `remove()` here, it seems unused?

Armaxis replied:
`toVisitLastSet.remove(removedNode.objectId)` uses it.

- [ ] Write unit tests for `LongScatterMap`

pyricau commented on shark-graph/src/main/java/shark/internal/hppc/LongScatterSet.kt:
```diff
@@ -192,4 +246,14 @@ class LongScatterSet {
+
+  companion object {
+
+    fun Set<Long>.toLongScatterSet(): LongScatterSet {
+      val longScatterSet = LongScatterSet()
+      longScatterSet.ensureCapacity(size)
+      forEach { longScatterSet.add(it) }
+      return longScatterSet
```
PathFinder is an internal class. You can change the signature of its methods and change everything else. Definitely shouldn't live here but in PathFinder if you think we should keep it, but I'm not sure why.

Armaxis replied:
- [ ] Move method to PathFinder

Armaxis commented on shark-graph/src/main/java/shark/internal/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+ * Replacement to a traditional Pair<Long, Object> that doesn't box long.
+ * Class is purposefully not made as a `data class` to decrease memory footprint of the object.
+ */
+class LongPair<out B>(
```
@pyricau I think we should rather rename the whole class to `LongObjectPair` so it is symmetrical to `LongObjectScattepMap` and `LongLongPair`

pyricau commented on shark-graph/src/main/java/shark/internal/hppc/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+
+/**
+ * Replacement to a traditional Pair<Long, Object> that doesn't box long.
+ * Class is purposefully not made as a `data class` to decrease memory footprint of the object.
```
> not made as a `data class`

=> "not a data class" ?

pyricau replied:
> to decrease memory footprint of the object

Can you outline why a data class would have a larger memory footprint in the javadoc comment?

pyricau commented on shark-graph/src/main/java/shark/internal/hppc/Tuples.kt:
```diff
@@ -0,0 +1,69 @@
+}
+
+/**
+ * Replacement to a traditional Pair<Long, Long> that doesn't box long.
```
> Replacement to a traditional Pair<Long, Long> 

=> Alternative to Pair<Long, Long>

Armaxis commented:
- Addressed all the pending comments
- Converted LongPair to `data class`
- Added Unit tests for LongScatterSet, 100% coverage. Disabled OutOfMemory test because it's slow.
![image](https://user-images.githubusercontent.com/2152942/87864665-ab6ecd00-c930-11ea-9900-9cfa564765a1.png)

Armaxis commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  @Ignore(value = "Takes ~20 seconds on MacBook Pro 2019, disabled so it doesn't waste CI time.")
+  @Test(expected = RuntimeException::class)
```
Alternatively we can check current free memory, fill it almost completely with an array and then start filling the set - it will be faster.
But I generally don't like an idea of throwing an OOM on CI in the test..

pyricau replied:
Why do we need to test this? Consider removing.

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,131 @@
+    set.release()
+
+    assertFalse(42 in set)
+    assertEquals(0, set.size())
```
we use assertJ everywhere else in the repo, let's stick to that.

Armaxis replied:
Good catch, thanks!

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,131 @@
+
+  @Test fun `verify add operation`() {
+    val set = LongScatterSet()
+    verifyAddOperation(set)
```
This test doesn't express what its testing, ie "verify add operation" doesn't tell me anything. Same for verifyRemoveOperation

pyricau replied:
@Armaxis same as before, can you provide a better name?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,131 @@
+  @Test fun `verify ensureCapacity does not break operations`() {
+    // Ensure capacity before adding/removing elements
+    verifyAddOperation(LongScatterSet().apply { ensureCapacity(10) })
+    verifyRemoveOperation(LongScatterSet().apply { ensureCapacity(10) })
```
This is the "arrange" part of the test. There shouldn't be any verification here, only set up. Asserts should stick to the last phase of a test.

Armaxis replied:
This is the actual testing part; but I can rewrite it to be cleaner

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,131 @@
+    set.add(testValues.first())
+    assertEquals(currentSize, set.size())
+  }
+}
```
newline?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,128 @@
+
+    // Ensure capacity before adding/removing elements
+    verifyAddOperation(setForAdditionOperationCheck)
+    verifyRemoveOperation(setForRemoveOperationCheck)
```
Same comment as before: this is the "test" part, so there shouldn't be any assertion or verification in it.

Armaxis replied:
Updated!

pyricau reviewed (approved):
Overall looking good but would you mind restructuting the tests a bit?

Armaxis commented:
How do you want them to be? Less shared code -> more obvious configuration/assertion parts?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -0,0 +1,142 @@
+
+    // First, verify adding unique values, as well as values that have matching hashKeys
+    // Values 11 and 14723950898 have same hash when it's calculated via HHPC.mixPhi()
+    Assertions.assertThat(HHPC.mixPhi(14723950898)).isEqualTo(HHPC.mixPhi(11))
```
nit: static imports

pyricau commented on shark-graph/src/test/java/shark/HhpcTest.kt:
```diff
@@ -0,0 +1,13 @@
+
+class HhpcTest {
+
+  @Test fun `11 and 14_723_950_898 has same hash`() {
```
nit: `has` => `have` no?

pyricau replied:
Might want to add a comment explaining why this is important?

pyricau replied:
Consider moving this test next to the tests where those values are used to make it clear what this is about.

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
+    TEST_VALUES_LIST.forEach { set.add(it) }
+
+    assertThat(set)
+        .contains(*TEST_VALUES_LIST.toLongArray())
```
Is there really only a `contains` that takes a vararg?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
+
+    assertThat(set)
+        .contains(*TEST_VALUES_LIST.toLongArray())
+        .hasSize(TEST_VALUES_LIST.size)
```
how about `containsExactly()` instead of those 2 asserts?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-    set.add(42)
-    set.release()
+
+    set.add(TEST_VALUE)
```
the name of the test says "+=" but the test calls add directly?

pyricau replied:
also I'm not sure what value exactly we get from testing this?

Armaxis replied:
Verification that we didn't accidentally break the operator. It's trivial, but still..

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-  @Test fun `verify LongScatterSet#ensureCapacity does not break Set when called before those operations`() {
-    val setForAddOperation = LongScatterSet().apply { ensureCapacity(10) }
-    val setForRemoveOperation = LongScatterSet().apply { ensureCapacity(10) }
+  @Test fun `adding element twice doesn't cause duplicates`() {
```
nit: "doesn't cause duplicates" => "is only added once" ?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-      add(10)
-      ensureCapacity(100)
-    }
+  @Test fun `adding elements with matching hash`() {
```
maybe "elements with equal hash can be added" ? the test name here says we're adding but doesn't outline the expected result.

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-    }
+    assertThat(set)
+        .doesNotContain(*TEST_VALUES_LIST.toLongArray())
+        .isEmpty()
```
is empty is enough.

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
 
-  private fun runAssertionsForRemoveOperation(set: LongScatterSet) {
-    require(set.size() == 0) { "Set should be empty in order to run the assertions on it" }
+  @Test fun `removing elements`() {
```
`LongScatterSet#remove() removes elements`

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
 
-    // Add test values to the set
-    testValues.forEach { set += it }
+  @Test fun `removing from empty set`() {
```
Is there any world in which a change to LongScatterSet could break this behavior?

Armaxis replied:
Of course! Any breaking change to `hashKey` or misconfiguration of initial values of `keys` or `mask` can cause incorrect result for `existing` (or throw an error)

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-        assertThat(set).contains(testValues[i])
-      }
-    }
+  @Test fun `removing elements with matching hash`() {
```
Probably want to test removing only one and verify that only the other is left inside the set?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-        .doesNotContain(42)
+  }
+
+  @Test fun `release operation cleans set`() {
```
`LongScatterSet#release() empties set`

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
 
-  private fun runAssertionsForAddOperation(set: LongScatterSet) {
-    require(set.size() == 0) { "Set should be empty in order to run the assertions on it" }
+  @Test fun `setting initial capacity before operations`() {
```
What is this testing exactly? Not obvious at all from test name and test content.

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-    }
+    assertThat(set)
+        .contains(TEST_VALUE)
+        .hasSize(1)
```
`containsExactly()` ?

pyricau commented on shark-graph/src/test/java/shark/LongScatterSetTest.kt:
```diff
@@ -1,142 +1,137 @@
-    // Verify that += operator works exactly like add()
-    set += 30
-    assertThat(set).contains(30)
+  @Test fun `adding a lot of elements causes resizing`() {
```
What's "a lot" ? How does the test demonstrate that resizing occurred? Also, resizing is not an interesting part the API, it's an impl detail, what should we test instead?

Armaxis replied:
this is a whitebox test - it makes sure that resizing doesn't damage the data. That's implementation details, you're right, but i want to test this scenario as well.

I can recycle it into "Stress test" and test set with million (billion?) addition operations

pyricau reviewed (commented):
thx! more nits!
